test/net.cpp

```diff
@@ -0,0 +1,207 @@
+	
+	KeyPair m_alias;
+	shared_ptr<TestNodeTable> nodeTable;
+	std::vector<std::pair<KeyPair,unsigned>> testNodes; // keypair and port
+	std::vector<shared_ptr<TestNodeTable>> nodeTables;
+};
+
+class TestUDPSocket: UDPSocketEvents, public TestHost
+{
+public:
+	TestUDPSocket(): m_socket(new UDPSocket<TestUDPSocket, 1024>(m_io, *this, 30300)) {}
+
+	void onDisconnected(UDPSocketFace*) {};
+	void onReceived(UDPSocketFace*, bi::udp::endpoint const&, bytesConstRef _packet) { if (_packet.toString() == "AAAA") success = true; }
+
+	shared_ptr<UDPSocket<TestUDPSocket, 1024>> m_socket;
+	
+	bool success = false;
+};
+
+BOOST_AUTO_TEST_CASE(test_neighbors_packet)
+{
+	KeyPair k = KeyPair::create();
+	std::vector<std::pair<KeyPair,unsigned>> testNodes(TestNodeTable::createTestNodes(16));
+	bi::udp::endpoint to(boost::asio::ip::address::from_string("127.0.0.1"), 30000);
+	
+	Neighbors out(to);
+	for (auto n: testNodes)
+	{
+		Neighbors::Node node;
+		node.ipAddress = boost::asio::ip::address::from_string("127.0.0.1").to_string();
+		node.port = n.second;
+		node.node = n.first.pub();
+		out.nodes.push_back(node);
+	}
+	out.sign(k.sec());
+
+	bytesConstRef packet(out.data.data(), out.data.size());
+	bytesConstRef rlpBytes(packet.cropped(97, packet.size() - 97));
+	Neighbors in = Neighbors::fromBytesConstRef(to, rlpBytes);
+	int count = 0;
+	for (auto n: in.nodes)
+	{
+		BOOST_REQUIRE_EQUAL(testNodes[count].second, n.port);
+		BOOST_REQUIRE_EQUAL(testNodes[count].first.pub(), n.node);
+		BOOST_REQUIRE_EQUAL(sha3(testNodes[count].first.pub()), sha3(n.node));
+		count++;
+	}
+}
+
+BOOST_AUTO_TEST_CASE(test_findnode_neighbors)
+{
+	// Executing findNode should result in a list which is serialized
+	// into Neighbors packet. Neighbors packet should then be deserialized
+	// into the same list of nearest nodes.
+}
+
+BOOST_AUTO_TEST_CASE(kademlia)
+{
+	TestNodeTableHost node(12);
+	node.start();
+	node.nodeTable->join(); // ideally, joining with empty node table logs warning we can check for
+	node.setup();
+	node.populate();
+	clog << "NodeTable:\n" << *node.nodeTable.get() << endl;
+	
+	node.pingAll();
+	this_thread::sleep_for(chrono::milliseconds(1000));
+	clog << "NodeTable:\n" << *node.nodeTable.get() << endl;
+	
+	node.nodeTable->reset();
+	clog << "NodeTable:\n" << *node.nodeTable.get() << endl;
+
+	node.populate(2);
+	this_thread::sleep_for(chrono::milliseconds(500));
+	clog << "NodeTable:\n" << *node.nodeTable.get() << endl;
+	
+	node.nodeTable->join();
+	this_thread::sleep_for(chrono::milliseconds(2000));
+	clog << "NodeTable:\n" << *node.nodeTable.get() << endl;
+}
+
+BOOST_AUTO_TEST_CASE(test_udp_once)
+{
+	UDPDatagram d(bi::udp::endpoint(boost::asio::ip::address::from_string("127.0.0.1"), 30300), bytes({65,65,65,65}));
+	TestUDPSocket a; a.m_socket->connect(); a.start();
+	a.m_socket->send(d);
+	sleep(1);
+	BOOST_REQUIRE_EQUAL(true, a.success);
+}
+
+BOOST_AUTO_TEST_SUITE_END()
+
```

Comment:
When I went to soapui.org the only thing I saw was advertisements trying to sell me certificates and training sessions for ~400eur and several links to wikipedia.

There aren't any 'optimization features' to remove. The optimizations mentioned in todos are not mandatory and they happen to be minor tasks.